vernac/metasyntax.ml

```diff
@@ -37,41 +37,90 @@ let pr_entry e =
   let () = Pcoq.Entry.print ft e in
   str (Buffer.contents entry_buf)
 
+let error_unknown_entry ?loc name =
+  user_err ?loc Pp.(str "Unknown or unprintable grammar entry " ++ str name ++ str".")
+
 let pr_registered_grammar name =
   let gram = Pcoq.find_grammars_by_name name in
   match gram with
-  | [] -> user_err Pp.(str "Unknown or unprintable grammar entry.")
+  | [] -> error_unknown_entry name
   | entries ->
-    let pr_one (Pcoq.AnyEntry e) =
+    let pr_one (Pcoq.Entry.Any e) =
       str "Entry " ++ str (Pcoq.Entry.name e) ++ str " is" ++ fnl () ++
       pr_entry e
     in
     prlist pr_one entries
 
-let pr_grammar = function
+let pr_grammar_subset grammar =
+  let pp = String.Map.mapi (fun name l -> match l with
+      | []  -> assert false
+      | entries ->
+        str "Entry " ++ str name ++ str " is" ++ fnl() ++
+        prlist_with_sep (fun () -> str "or" ++ fnl())
```

Comment:
There are a few cases or entries sharing a name. With -noinit:
~~~
Entry ident is
[ LEFTA
  [ IDENT ] ]
or
[ LEFTA
  [ ident ] ]

Entry test_lpar_id_coloneq is
[ <parser> ]
or
[ <parser> ]
~~~

With the prelude:
~~~
Entry binders is
[ LEFTA
  [ binders ] ]
or
[ LEFTA
  [ LIST0 binder ] ]

Entry ident is
[ LEFTA
  [ IDENT ] ]
or
[ LEFTA
  [ ident ] ]

Entry lpar_id_colon is
[ <parser> ]
or
[ <parser> ]

Entry lpar_id_coloneq is
[ <parser> ]
or
[ <parser> ]

Entry test_lpar_id_coloneq is
[ <parser> ]
or
[ <parser> ]
~~~

With Require Ltac2.Ltac2:
~~~
Entry binder is
[ LEFTA
  [ "_"
  | ident ] ]
or
[ LEFTA
  [ name
  | closed_binder ] ]

Entry binders is
[ LEFTA
  [ binders ] ]
or
[ LEFTA
  [ LIST0 binder ] ]

Entry branches is
[ LEFTA
  [ "|"; LIST1 branch SEP "|"
  | LIST1 branch SEP "|"
  |  ] ]
or
[ LEFTA
  [ OPT "|"; LIST0 eqn SEP "|" ] ]

Entry ident is
[ LEFTA
  [ IDENT ] ]
or
[ LEFTA
  [ ident ] ]

Entry input_fun is
[ LEFTA
  [ tac2pat LEVEL "0" ] ]
or
[ LEFTA
  [ "_"
  | ident ] ]

Entry let_clause is
[ LEFTA
  [ let_binder; type_cast; ":="; ltac2_expr ] ]
or
[ LEFTA
  [ [ "_" ]; ":="; ltac_expr
  | identref; ":="; ltac_expr
  | identref; LIST1 input_fun; ":="; ltac_expr ] ]

Entry lpar_id_colon is
[ <parser> ]
or
[ <parser> ]

Entry lpar_id_coloneq is
[ <parser> ]
or
[ <parser> ]

Entry tactic_atom is
[ LEFTA
  [ "@"; ident
  | "&"; lident
  | "'"; constr
  | IDENT "constr"; ":"; "("; lconstr; ")"
  | IDENT "open_constr"; ":"; "("; lconstr; ")"
  | IDENT "ident"; ":"; "("; lident; ")"
  | IDENT "pat"; ":"; "("; cpattern; ")"
  | IDENT "reference"; ":"; "("; globref; ")"
  | IDENT "ltac1"; ":"; "("; ltac1_expr_in_env; ")"
  | IDENT "ltac1val"; ":"; "("; ltac1_expr_in_env; ")"
  | integer
  | string
  | qualid ] ]
or
[ LEFTA
  [ "()"
  | integer
  | reference ] ]

Entry test_lpar_id_coloneq is
[ <parser> ]
or
[ <parser> ]
~~~

Reply:
My recollection is that the parser doesn't ensure unique names when you add productions through the programmatic interface and the rules are expressed in a way that doesn't rely on the nonterminal name.  Something like that.